confgenerator/confgenerator.go

```diff
@@ -97,6 +99,152 @@ func (m *Metrics) generateOtelPipelines() (map[string]otel.Pipeline, error) {
 	return out, nil
 }
 
+func (uc *UnifiedConfig) GenerateFluentBitConfigsModular(logsDir string, stateDir string, hostInfo *host.InfoStat) (string, string, error) {
+	userAgent, _ := getUserAgent("Google-Cloud-Ops-Agent-Logging", hostInfo)
+	components, err := uc.Logging.generateFluentbitComponents(userAgent, hostInfo)
+	if err != nil {
+		return "", "", err
+	}
+
+	c := fluentbit.ModularConfig{
+		Variables: map[string]string{
+			"buffers_dir": path.Join(stateDir, "buffers"),
+			"logs_dir":    logsDir,
+		},
+		Components: components,
+	}
+	return c.Generate()
+}
+
+type fbSource struct {
+	tag        string
+	components []fluentbit.Component
+}
+
+func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.InfoStat) ([]fluentbit.Component, error) {
+	var out []fluentbit.Component
+	out = append(out, fluentbit.Service{}.Component())
+
+	if l != nil && l.Service != nil {
+		var sources []fbSource
+		var logNames []string
+		for pID, p := range l.Service.Pipelines {
+			for _, rID := range p.ReceiverIDs {
+				receiver, ok := l.Receivers[rID]
+				if !ok {
+					return nil, fmt.Errorf("receiver %q not found", rID)
+				}
+				tag := fmt.Sprintf("%s.%s", pID, rID)
+				components := receiver.Components(tag)
+				for i, pID := range p.ProcessorIDs {
+					processor, ok := l.Processors[pID]
+					if !ok {
+						processor, ok = LegacyBuiltinProcessors[pID]
+					}
+					if !ok {
+						return nil, fmt.Errorf("processor %q not found", pID)
+					}
+					components = append(components, processor.Components(tag, i)...)
+				}
+				components = append(components, setLogNameComponents(tag, rID)...)
+				logNames = append(logNames, regexp.QuoteMeta(rID))
+				sources = append(sources, fbSource{tag, components})
+			}
+		}
+		sort.Slice(sources, func(i, j int) bool { return sources[i].tag < sources[j].tag })
+		sort.Strings(logNames)
+
+		for _, s := range sources {
+			out = append(out, s.components...)
+		}
+		if len(logNames) > 0 {
```

Comment:
Is this possible? Would this mean a `Logging` section with no pipelines, or a pipeline with no receivers? Do we want to support such a config? Would it make sense to instead add a validator to prevent this?

Reply:
This was needed to handle one of the test inputs. Therefore I believe for backwards compatibility reasons we can't reject it.

Reply:
It was indeed a pipeline with no receivers (`logging-default_overrides_disable_all/input.yaml`). SIgh.